### PR TITLE
replace ToUtf8 with Encoding.UTF8.GetBytes

### LINKED ISSUE #75779: JsonWriterHelper.Transcoding.cs replacement
old discussion: https://github.com/dotnet/corefx/pull/34425#discussion_r245931534

https://github.com/dotnet/runtime/blob/eeef4b14c5a137c2f665a9be4f30e4383d1038da/src/libraries/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Transcoding.cs#L28

the corelib implementation this `ToUtf8` method was shadowing back in 2018-2019 is accelerated with hardware intrinsics since and code was heavily refactored, but `JsonWriterHelper.Transcoding` is stuck in time. should it be replaced by corelib version?

### PR DESCRIPTION
**Pull request by kasperk81**
fixes #75779

### CODE REVIEW

**Review by stephentoub — COMMENTED** [review]
Thanks!

Can we validate performance is as we expect it to be (as good or better than it was before)?

**Comment by eiriktsarpalis** [comment]
> Can we validate performance is as we expect it to be (as good or better than it was before)?

@kasperk81 If you're interested, here are [a few instructions](https://github.com/dotnet/performance/blob/main/src/benchmarks/micro/README.md#private-runtime-builds) on how to benchmark private build performance. The STJ benchmarks can be run using the `System.Text.Json*` filter in the BenchmarkDotNet. I can also do it for you if you prefer.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Transcoding.cs [hunk 1]
```diff
@@ -1,339 +0,0 @@
...
-        /// <param name="utf16Source">A span containing a sequence of UTF-16 bytes.</param>
-        /// <param name="utf8Destination">A span to write the UTF-8 bytes into.</param>
-        /// <param name="bytesConsumed">On exit, contains the number of bytes that were consumed from the <paramref name="utf16Source"/>.</param>
-        /// <param name="bytesWritten">On exit, contains the number of bytes written to <paramref name="utf8Destination"/></param>
-        /// <returns>A <see cref="OperationStatus"/> value representing the state of the conversion.</returns>
-        public static unsafe OperationStatus ToUtf8(ReadOnlySpan<byte> utf16Source, Span<byte> utf8Destination, out int bytesConsumed, out int bytesWritten)
-        {
-            //
-            //
-            // KEEP THIS IMPLEMENTATION IN SYNC WITH https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Text/UTF8Encoding.cs#L845
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Transcoding.cs:32]: I actually wasn't aware we had this method until you brought it up, and I must say I'm glad to see it go. Thanks :-)

File: src/libraries/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs [hunk 2]
```diff
@@ -531,21 +531,8 @@ private bool TrySkipHelper()
                 otherUtf8Text = stackalloc byte[JsonConstants.StackallocByteThreshold];
             }
 
-            ReadOnlySpan<byte> utf16Text = MemoryMarshal.AsBytes(text);
-            OperationStatus status = JsonWriterHelper.ToUtf8(utf16Text, otherUtf8Text, out int consumed, out int written);
-            Debug.Assert(status != OperationStatus.DestinationTooSmall);
-            bool result;
-            if (status > OperationStatus.DestinationTooSmall)   // Equivalent to: (status == NeedMoreData || status == InvalidData)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs:538]: Case in point: if I'm understanding this code correctly, this method will now throw when in the past it was returning `false`.

**Comment by kasperk81** [comment]
> Can we validate performance is as we expect it to be (as good or better than it was before)?

from one set of performance tests, i'm getting mixed results.

summary:
better: 5, geomean: 1.087
worse: 4, geomean: 1.115
total diff: 9

| Slower                                                                           | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| -------- |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf8(Formatted: False, SkipValidatio |      1.16 |           754.87 |           876.72 |         |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf16(Formatted: False, SkipValidati |      1.15 |           863.18 |           992.44 | bimodal |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf8(Formatted: False, SkipValidatio |      1.12 |       1993853.75 |       222316.66 |         |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf8(Formatted: False, SkipValidatio |      1.04 |       1911530.87 |       1982514.57 |         |

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf16(Formatted: False, SkipValidati |      1.15 |           832.79 |           721.12 |         |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf16(Formatted: True, SkipValidatio |      1.08 |          1063.42 |           980.44 |         |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf16(Formatted: True, SkipValidatio |      1.08 |          1019.21 |           940.08 |         |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf8(Formatted: True, SkipValidation |      1.07 |          1003.61 |           941.22 |         |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf8(Formatted: False, SkipValidatio |      1.05 |           710.38 |           676.76 |         |

**Comment by EgorBo** [comment]
@kasperk81 I'd recommend to pass `--envVars DOTNET_ReadyToRun:0` to BDN to disable static PGO as your changes changed layout and made the built-in profile stale.
I assume we need to reflect it in the docs (https://github.com/dotnet/performance/blob/main/src/benchmarks/micro/README.md#private-runtime-builds)

**Comment by kasperk81** [comment]
> I'd recommend to pass `--envVars DOTNET_ReadyToRun:0`

with envVars, it's clear that diff is slower
```
summary:
worse: 10, geomean: 1.132
total diff: 10

| Slower                                                                           | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| -------- |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf16(Formatted: True, SkipValidatio |      1.29 |       2362864.09 |       3058230.01 |         |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf8(Formatted: True, SkipValidation |      1.22 |       2414225.03 |       2944031.19 |         |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf16(Formatted: False, SkipValidati |      1.21 |           689.49 |           834.19 |         |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf16(Formatted: False, SkipValidati |      1.14 |           748.64 |           856.18 |         |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf16(Formatted: True, SkipValidatio |      1.12 |           963.79 |          1079.45 | bimodal |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf16(Formatted: True, SkipValidatio |      1.10 |           908.93 |           995.58 |         |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf16(Formatted: False, SkipValidati |      1.07 |       1838042.24 |       1969624.88 |         |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf8(Formatted: False, SkipValidatio |      1.07 |           634.40 |           677.76 |         |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf8(Formatted: True, SkipValidation |      1.06 |           893.44 |           947.61 |         |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf8(Formatted: False, SkipValidatio |      1.06 |           707.75 |           750.28 |         |

No Faster results for the provided threshold = 2% and noise filter = 0.3 ns.
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.String.cs [hunk 3]
```diff
@@ -1380,14 +1382,16 @@ private void WriteStringMinimized(ReadOnlySpan<char> escapedPropertyName, ReadOn
...
+            JsonWriterHelper.TryGetUtf8FromText(escapedPropertyName, output.Slice(BytesPending), out int written);
+            BytesPending += written;
 
             output[BytesPending++] = JsonConstants.Quote;
             output[BytesPending++] = JsonConstants.KeyValueSeparator;
 
             output[BytesPending++] = JsonConstants.Quote;
 
-            TranscodeAndWrite(escapedValue, output);
+            JsonWriterHelper.TryGetUtf8FromText(escapedPropertyName, output.Slice(BytesPending), out written);
```
Comment (kasperk81) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.String.cs:1393]:
```suggestion
            JsonWriterHelper.TryGetUtf8FromText(escapedValue, output.Slice(BytesPending), out written);
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs [hunk 4]
```diff
@@ -229,5 +229,35 @@ internal static void ValidateNumber(ReadOnlySpan<byte> utf8FormattedNumber)
                     nameof(utf8FormattedNumber));
             }
         }
+
+        public static readonly UTF8Encoding s_utf8Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false);
```
Comment (jkotas) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs:233]: Catching UTF8Encoding like this is slight de-optimization. ` Encoding.UTF8` has same cache, and it also returns internal more specialized sealed `UTF8Encoding` type that allows de-virtualization to kick-in in more situations.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs [hunk 5]
```diff
@@ -229,5 +229,35 @@ internal static void ValidateNumber(ReadOnlySpan<byte> utf8FormattedNumber)
...
+#else
+                if (text.IsEmpty)
+                {
+                    return true;
+                }
+
+                fixed (char* charPtr = text)
+                fixed (byte* destPtr = dest)
+                {
+                    written = s_utf8Encoding.GetBytes(charPtr, text.Length, destPtr, dest.Length);
```
Comment (kasperk81) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs:253]:
```suggestion
        internal static unsafe bool TryGetUtf8FromText(ReadOnlySpan<char> text, Span<byte> dest, out int written)
        {
            written = 0;

            try
            {
#if NETCOREAPP
                written = Encoding.UTF8.GetBytes(text, dest);
                return true;
#else
                if (text.IsEmpty)
                {
                    return true;
                }

                fixed (char* charPtr = text)
                fixed (byte* destPtr = dest)
                {
                    written = Encoding.UTF8.GetBytes(charPtr, text.Length, destPtr, dest.Length);
```

**Comment by eiriktsarpalis** [comment]
Change seems to be introducing a couple of test failures, see:

https://github.com/dotnet/runtime/pull/75834/checks?check_run_id=8439815257

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs [hunk 6]
```diff
@@ -229,5 +229,33 @@ internal static void ValidateNumber(ReadOnlySpan<byte> utf8FormattedNumber)
...
+                {
+                    return true;
+                }
+
+                fixed (char* charPtr = text)
+                fixed (byte* destPtr = dest)
+                {
+                    written = Encoding.UTF8.GetBytes(charPtr, text.Length, destPtr, dest.Length);
+                    return true;
+                }
```
Comment (kasperk81) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs:253]:
```suggestion
        private static readonly UTF8Encoding s_utf8Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);

        internal static unsafe bool TryGetUtf8FromText(ReadOnlySpan<char> text, Span<byte> dest, out int written)
        {
            written = 0;

            try
            {
#if NETCOREAPP
                written = s_utf8Encoding.GetBytes(text, dest);
                return true;
#else
                if (text.IsEmpty)
                {
                    return true;
                }

                fixed (char* charPtr = text)
                fixed (byte* destPtr = dest)
                {
                    written = s_utf8Encoding.GetBytes(charPtr, text.Length, destPtr, dest.Length);
                    return true;
                }
```

**Comment by eiriktsarpalis** [comment]
I've analyzed the test failures and made the following discoveries. Unlike other strings passed to `Utf8JsonWriter`, comment strings are not escaped, by design:

https://github.com/dotnet/runtime/blob/58a1180ad6421c6f69cf307b57b2496c888585f2/src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Comment.cs#L45

As a result, the comment writer supports inputs containing unpaired surrogates:

https://github.com/dotnet/runtime/blob/58a1180ad6421c6f69cf307b57b2496c888585f2/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs#L4165

However, in such cases the current implementation will simply write the comment string up to the impacted character and discard everything else

https://github.com/dotnet/runtime/blob/58a1180ad6421c6f69cf307b57b2496c888585f2/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs#L4200

Even though we're testing for this, I don't think anybody could reasonably depend on the current behavior. I would recommend taking a breaking change and have the method throw ArgumentException instead.

**Comment by kasperk81** [comment]
string before write: `"comment is / * valid even with unpaired surrogate \udc00 this part no longer visible"`
expected: `"comment is / * valid even with unpaired surrogate`
actual: `comment is /* valid*//**//**/`

actual doesn't make sense, it replaced 30 expected characters with `*//**//**/` and eats the space between forward slash and asterisk in `/ * valid` part.

Encoding.UTF8.GetBytes is lacking something that ToUtf8 was handling well (that's the only thing i've replaced). could be an indication of bug in GetBytes that goes beyond this odd comment?

if i delete that comment from the test, it passes though the other comments like `"comment is /* valid"` are still there
```
--- a/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs
+++ b/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs
@@ -4162,10 +4162,6 @@ public void WriteCommentsInvalidTextAllowed(bool formatted, bool skipValidation)
             jsonUtf8.WriteCommentValue(comment.AsSpan());
             jsonUtf8.WriteCommentValue(Encoding.UTF8.GetBytes(comment));
 
-            comment = "comment is / * valid even with unpaired surrogate \udc00 this part no longer visible";
-            jsonUtf8.WriteCommentValue(comment);
-            jsonUtf8.WriteCommentValue(comment.AsSpan());
-
             jsonUtf8.Flush();
 
             // Explicitly skipping flushing here
@@ -4197,10 +4193,6 @@ private static string GetCommentExpectedString(bool prettyPrint)
             json.WriteComment(comment);
             json.WriteComment(comment);
 
-            comment = "comment is / * valid even with unpaired surrogate ";
-            json.WriteComment(comment);
-            json.WriteComment(comment);
-
             json.Flush();
 
             return Encoding.UTF8.GetString(ms.ToArray());
```

**Pull request closed without merging**